File: kotlin/Hash Maps and Sets/PairSumUnSortedTwoPass.kt
fun pairSumUnsortedTwoPass(nums: List<Int>, target: Int): List<Int> {
    val numMap = mutableMapOf<Int, Int>()
    // First pass: Populate the hash map with each number and its 
    // index.
    for ((i, num) in nums.withIndex()) {
        numMap[num] = i
    }
    // Second pass: Check for each number's complement in the hash map.
    for ((i, num) in nums.withIndex()) {
        val complement = target - num
        if (complement in numMap && numMap[complement] != i) {
            return listOf(i, numMap[complement]!!)
        }
    }
    return emptyList()
}

File: typescript/Hash Maps and Sets/pair_sum_unsorted_two_pass.ts
function pairSumUnsortedTwoPass(nums: number[], target: number): number[] {
    const numMap: { [key: number]: number } = {};
    // First pass: Populate the hash map with each number and its 
    // index.
    for (let i = 0; i < nums.length; i++)
        numMap[nums[i]] = i;
    // Second pass: Check for each number's complement in the hash map.
    for (let i = 0; i < nums.length; i++) {
        const complement = target - nums[i];
        if (numMap[complement] !== undefined && numMap[complement] !== i) 
            return [i, numMap[complement]];
    }
    return [];
}
